pkg/util/k8s.go

```diff
@@ -0,0 +1,31 @@
+package util
+
+import (
+	"fmt"
+	"strings"
+)
+
+// EncodeKubernetesName takes a string and turns it into a form suitable for use as a Kubernetes name.
+// K8S names are lower case, numbers, '-', and '.'. Invalid characters are percent-encoded in a _similar_ way to URL
+// encoding, with a percent followed by the numerical character code. Conversion is:
+//   Upper case letters -> lower case.
+//   'a-z', '0-9', '-', '.' -> left as-is.
+//   Any other characters: percent-encoded ('%' + rune hex code).
+func EncodeKubernetesName(name string) string {
+	name = strings.ToLower(name)
```

Comment:
what about collisions?
I can have `org/repo/branch` and `org/repo/BrAnCh` and they are different things.

Reply:
Good point, that _was_ the original point of this change, after all!
Upper case are now converted to lower case (to keep things as readable as possible), and given a prefix to avoid upper/lower collisions like this.
I also realised another key requirement hadn't been met with the original encoding, as % isn't a valid character in a K8S name. So the encoding is as-before with codes, but now uses period in place of percent.